src/analyses/ai_history.h

```diff
@@ -210,6 +210,67 @@ class ahistoricalt : public ai_history_baset
   }
 };
 
+/// Records the call stack
+/// Care must be taken when using this on recursive code; it will need the
+/// domain to be capable of limiting the recursion.
+class call_stack_historyt : public ai_history_baset
+{
+protected:
+  class call_stack_entryt;
+  typedef std::shared_ptr<const call_stack_entryt> cse_ptrt;
+  class call_stack_entryt
+  {
+  public:
+    locationt current_location;
+    cse_ptrt caller;
+
+    call_stack_entryt(locationt l, cse_ptrt p) : current_location(l), caller(p)
+    {
+    }
+
+    bool operator<(const call_stack_entryt &op) const;
+    bool operator==(const call_stack_entryt &op) const;
+  };
+
+  cse_ptrt current_stack;
+  // DATA_INVARIANT(current_stack != nullptr, "current_stack must exist");
+  // DATA_INVARIANT(current_stack->current.is_dereferenceable(),
+  //                "Must not be _::end()")
+
+  explicit call_stack_historyt(cse_ptrt p)
+    : ai_history_baset(p->current_location), current_stack(p)
+  {
+    PRECONDITION(p != nullptr); // A little late by now but worth documenting
+  }
+
+public:
+  explicit call_stack_historyt(locationt l)
+    : ai_history_baset(l), current_stack(new call_stack_entryt(l, nullptr))
+  {
+  }
+
+  call_stack_historyt(const call_stack_historyt &old)
+    : ai_history_baset(old), current_stack(old.current_stack)
+  {
+  }
+
+  step_returnt step(locationt to, const trace_sett &others) const override;
+
+  bool operator<(const ai_history_baset &op) const override;
+  bool operator==(const ai_history_baset &op) const override;
+
+  const locationt &current_location(void) const override
+  {
+    return current_stack->current_location;
+  }
+
+  // Use default widening
+  // Typically this would be used for loops, which are not tracked
+  // it would be possible to use this to improve the handling of recursion
```

Comment:
Rather than just a comment saying that the default implementation is used, I still prefer to explicitly define the function and invoke the parent.

Reply:
Done!